IdentityCore/src/throttling/model/MSIDThrottlingModelInteractionRequire.m

```diff
@@ -0,0 +1,123 @@
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.  
+
+
+#import <Foundation/Foundation.h>
+#import "MSIDThrottlingModelInteractionRequire.h"
+#import "MSIDThrottlingMetaData.h"
+#import "MSIDThrottlingMetaDataCache.h"
+#import "NSDate+MSIDExtensions.h"
+#import "NSError+MSIDExtensions.h"
+static NSInteger const DefaultUIRequired = 120;
+
+@implementation MSIDThrottlingModelInteractionRequire
+
+- (instancetype) initWithRequest:(id<MSIDThumbprintCalculatable> _Nonnull)request
+                     cacheRecord:(MSIDThrottlingCacheRecord * _Nullable)cacheRecord
+                   errorResponse:(NSError *)errorResponse
+                     accessGroup:(NSString *)accessGroup
+{
+    self = [super init];
+    if (self)
+    {
+        self.thumbprintType = MSIDThrottlingThumbprintTypeFull;
+        self.thumbprintValue = [request fullRequestThumbprint];
+        self.throttleDuration = DefaultUIRequired;
+    }
+    return self;
+}
+
+
+/**
+ * if is appliable for UIRequired:
+ * Throttle conditions:
+ * MSAL <-> server flow: OAuth error: error.msidOauthError is not nil
+ * MSAL <-> SSO-Ext <-> Server flow: error.code MSALErrorInteractionRequired
+ */
++ (BOOL)isApplicableForTheThrottleModel:(NSError *)errorResponse
+{
+    // MSALErrorInteractionRequired = -50002
+    NSSet *uirequiredErrors = [[NSSet alloc] initWithArray:@[[NSNumber numberWithInt:(-50002)]]];
+    BOOL isMSIDError = [errorResponse.domain hasPrefix:@"MSID"];
+    
+    if (isMSIDError)
+    {
+        NSString *errorString = errorResponse.msidOauthError;
+        NSUInteger errorCode = errorResponse.code;
+        if ([NSString msidIsStringNilOrBlank:errorString] || (errorCode == MSIDErrorInteractionRequired))
+        {
+            return YES;
+        }
+    }
+    else
+    {
+        if ([uirequiredErrors containsObject:[NSNumber numberWithLong:errorResponse.code]])
+        {
+            return YES;
+        }
+    }
+    return NO;
+}
+
+- (BOOL)shouldThrottleRequest
+{
+    BOOL res = NO;
+    NSError *error;
+    NSDate *currentTime = [NSDate date];
+    NSDate *lastRefreshTime = [MSIDThrottlingModelInteractionRequire getLastRefreshTimeAccessGroup:self.accessGroup context:self.context error:&error];
+    // If currentTime is later than the expiration Time or the lastRefreshTime is later then the expiration Time, we don't throttle the request
+    if ([currentTime compare:self.cacheRecord.expirationTime] != NSOrderedAscending
+        || (lastRefreshTime && [lastRefreshTime compare:self.cacheRecord.expirationTime] != NSOrderedAscending))
+    {
+        [self.cacheService removeObjectForKey:self.thumbprintValue error:&error];
+        if (error)
+        {
+            MSID_LOG_WITH_CTX(MSIDLogLevelError, self.context, @"Throttling: error when remove record from database %@ ", error);
+        }
+        res = NO;
```

Comment:
nit: I think you can just return NO & do away with res variable.

Reply:
Addressed.